README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ActionSquirrel

<!-- badges: start -->

[![Project Status: Concept – Minimal or no implementation has been done
yet, or the repository is only intended to be a limited example, demo,
or
proof-of-concept.](https://www.repostatus.org/badges/latest/concept.svg)](https://www.repostatus.org/#concept)
[![Blog
post](https://img.shields.io/badge/rostrum.blog-post-008900?labelColor=000000&logo=data%3Aimage%2Fgif%3Bbase64%2CR0lGODlhEAAQAPEAAAAAABWCBAAAAAAAACH5BAlkAAIAIf8LTkVUU0NBUEUyLjADAQAAACwAAAAAEAAQAAAC55QkISIiEoQQQgghRBBCiCAIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAAh%2BQQJZAACACwAAAAAEAAQAAAC55QkIiESIoQQQgghhAhCBCEIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAA7)](https://www.rostrum.blog/2021/10/03/squirrel/)
<!-- badges: end -->

{ActionSquirrel} is an R package containing an interactive audio-visual
experience (‘game’) in the R console, powered by
[{R6}](https://r6.r-lib.org/).

You play as the titular Action Squirrel. Move around the forest and
collect at least eight nuts to survive winter, which happens after 30
turns. Avoid being eaten by the owl.

Read more in [the accompanying blog
post](https://www.rostrum.blog/2021/10/03/squirrel/).

## How to play

### Install

The package is available from GitHub.

``` r
install.packages("remotes")  # if not installed already
remotes::install_github("matt-dray/ActionSquirrel")
```

Note that I’ve developed and tested this only in RStudio v1.4.1717 with
R v4.1.0 running on macOS Big Sur.

Attach the package for instructions.

``` r
library(ActionSquirrel)
# Welcome to {ActionSquirrel}!
# * New game: x <- ActionSquirrel$new()
# * Move:     e.g. x$move('up')
# * Info:     x$pause()
```

### Start

Initiate by assigning `ActionSquirrel$new()` to a name. This clears the
console and generates a forest that contains Action Squirrel. And a
delicious nut. And a scary owl!

``` r
x <- ActionSquirrel$new()
# 🌳 🌳 🌳 🌳 🌳 
# 🌳 🌳 🌰 🌳 🌳 
# 🌳 🌳 🌳 🐿 🌳 
# 🌳 🦉 🌳 🌳 🌳 
# 🌳 🌳 🌳 🌳 🌳 
# Moves: 0 
# Nuts: 0
```

Note: Action Squirrel nor the nut nor the owl are as big as a tree. It’s
just perspective.

### Move

You control Action Squirrel. Use the `move()` method to move through the
forest and grab the nuts, avoiding the owl.

``` r
x$move("up")
# 🌳 🌳 🌳 🌳 🌳 
# 🌳 🌳 🌰 🐿️ 🌳 
# 🌳 🦉 🌳 🌳 🌳 
# 🌳 🌳 🌳 🌳 🌳 
# 🌳 🌳 🌳 🌳 🌳 
# Moves: 1 
# Nuts: 0
```

Congratulations, your move tally has increased by one. Winter arrives
after 30 turns and the game will end.

To the nut!

``` r
x$move("left")
# 🌳 🌳 🌳 🌳 🌳 
# 🌳 🌳 🐿️ 🌳 🌳 
# 🌳 🦉 🌳 🌳 🌳 
# 🌰 🌳 🌳 🌳 🌳 
# 🌳 🌳 🌳 🌳 🌳 
# Moves: 2 
# Nuts: 1
```

Congratulations, your nut tally has increased by one. And there’s a new
nut! Collect at least eight, or you won’t survive winter.

Advanced tech: you only need the first letter of the direction, like
`x$move("r")` for right, and you can chain moves together like
`x$move("r")$move("d")` to go right then down.

I’ll leave it to the player to discover the consequences of being eaten,
the endscreens and the sound effects.

## Thanks

Inspired in part by [Tomaz’s ‘Little Useless-Useful R
Functions’](https://tomaztsql.wordpress.com/2021/09/13/little-useless-useful-r-functions-year-progress-bar/).

Thanks to Winston Chang for [the {R6} package](https://r6.r-lib.org/),
and Stefan Siegert and Robin Williams for
[{sonify}](https://CRAN.R-project.org/package=sonify).

## Similar projects

I’ve done some other silly things with {R6}. You can:

-   [play a text-based version of *Pokemon Blue*’s Safari Zone in
    R](https://www.rostrum.blog/2021/01/04/safar6/)
-   [repay your loan to *Animal Crossing*’s Tom Nook with an Automatic
    Bell
    Dispenser](https://www.rostrum.blog/2020/04/04/repaying-tom-nook-with-r6/)

## Code of Conduct

Add bugs or ideas to [the
issues](https://github.com/matt-dray/ActionSquirrel/issues).

Please note that the {ActionSquirrel} project is released with a
[Contributor Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# ActionSquirrel 0.1.0

* Added blog link to README, bump 0.1.0.

# ActionSquirrel 0.0.0.9007

* Added nut cache win/lose scenarios (#14).

# ActionSquirrel 0.0.0.9006

* Added 'active' state, preventing user from playing beyond a game over (#12).
* Removed `print()` method, added `pause()` method.

# ActionSquirrel 0.0.0.9005

* Allowed chaining of methods.
* Created basic `print` method.

# ActionSquirrel 0.0.0.9004

* Added owl AI (#3).
* Added death outcome.
* Added game 'end' (#4).

# ActionSquirrel 0.0.0.9003

* Added sound effects for movement and nut capture (#9).
* Confined `move` method direction options (#6).

# ActionSquirrel 0.0.0.9002

* Added logic to place and intercept nuts, and increase nut tally (#2).

# ActionSquirrel 0.0.0.9001

* Added logic to warn player about reaching the edge (#1).

# ActionSquirrel 0.0.0.9000

* Set up package with DESCRIPTION, README, license, CoC, NEWS, .gitignore.
* Added class.R, zzz.R.


R/class.R

#' Generate An R6-Class 'Action Squirrel' Game
#'
#' Prints to the console a grid of emoji representing an overworld,
#' including a player character that can be controlled by the user.
#' Also provides a live tally of moves taken.
#'
#' @return Console output with \code{cat()}.
#' @export
#'
#' @examples
#'     \dontrun{
#'     x <- ActionSquirrel$new()  # generate overworld
#'     x$move("up")               # move character up
#'     x$move("right")            # move character right
#'     }
ActionSquirrel <- R6::R6Class(

  "ActionSquirrel",

  public = list(

    # FIELDS ----

    #' @field active Logcial. Is the game in play?
    active = TRUE,

    #' @field overworld Character vector. The overworld in one dimension.
    overworld = rep("\U1F333", 25),

    #' @field s_loc Numeric. The location of the squirrel.
    s_loc = 14,

    #' @field n_loc Numeric. The location of the nut.
    n_loc = 8,

    #' @field o_loc Numeric. The location of the owl.
    o_loc = 17,

    #' @field moves Numeric. The number of moves made by the player.
    moves = 0,

    #' @field nuts Numeric. The number of nuts collected by the player.
    nuts = 0,

    # METHODS ----

    #' @description Create a new overworld.
    #' @return An R6-class object.
    initialize = function() {

      self$overworld[self$n_loc] <- "\U1F330"
      self$overworld[self$s_loc] <- "\U1F43F"
      self$overworld[self$o_loc] <- "\U1F989"
      overworld_mat <- t(matrix(self$overworld, nrow = 5))
      cat("\014")
      for (i in seq(nrow(overworld_mat))) {
        cat(overworld_mat[i, ], "\n")
      }
      cat("Moves:", self$moves, "\nNuts:", self$nuts)

    },

    #' @description Pause menu with instructions.
    #' @return An R6-class object.
    pause = function() {
      cat(
        "P A U S E\n",
        "* Aim:       get eight nuts before winter (30 moves)\n",
        "* Move:      e.g. x$move('up')\n",
        "* Chain:     e.g. x$move('u')$move('r')\n",
        "* New game:  x <- ActionSquirrel$new()\n",
        "* Info:      x$pause()\n",
        "* Source:    github.com/matt-dray/ActionSquirrel"
      )
      invisible(self)
    },

    #' @description Move the hero around the overworld.
    #' @param where Character. Which direction to move. One of \code{"up"},
    #'      \code{"down"}, \code{"left"} and \code{"right"}.
    #' @return An R6-class object.
    move = function(where = c("up", "down", "left", "right")) {

      where <- match.arg(where)

      # If game is in play
      if (self$active) {

        # Stop at grid edge
        if (
          (where == "up" & self$s_loc %in% 1:5) |
          (where == "down" & self$s_loc %in% 21:25) |
          (where == "left" & self$s_loc %in% seq(1, 25, 5)) |
          (where == "right" & self$s_loc %in% seq(5, 25, 5))
        ) {

          sonify::sonify(1, 1, duration = 0.001)
          sonify::sonify(1, 1, duration = 0.001)
          return(cat("You reached the edge.\nTry another direction."))

        }

        # Increment grid location
        if (where == "up")    { move_n <- -5}
        if (where == "down")  { move_n <- +5}
        if (where == "left")  { move_n <- -1}
        if (where == "right") { move_n <- +1}
        self$s_loc <- self$s_loc + move_n

        # Move sound
        if (self$s_loc != self$n_loc) {
          sonify::sonify(1, 1, duration = 0.01)
        }

        # Nut capture routine
        if (self$s_loc == self$n_loc) {
          sonify::sonify(c(1, 0), c(1, 0), duration = 0.1)
          self$nuts <- self$nuts + 1 # increment nut tally
          self$n_loc <- sample(seq(25)[-self$s_loc], 1)  # new nut location
        }

        # Owl move
        # corners
        if (self$o_loc == 1)  { o_move <- sample(c(0,  1,  5), 1) }
        if (self$o_loc == 5)  { o_move <- sample(c(0, -1,  5), 1) }
        if (self$o_loc == 21) { o_move <- sample(c(0,  1, -5), 1) }
        if (self$o_loc == 25) { o_move <- sample(c(0, -1, -5), 1) }
        # edges
        if (self$o_loc %in% 2:4)           { o_move <- sample(c(0, -1,  1,  5), 1) }
        if (self$o_loc %in% c(6, 11, 16))  { o_move <- sample(c(0,  1, -5,  5), 1) }
        if (self$o_loc %in% c(10, 15, 20)) { o_move <- sample(c(0, -1, -5,  5), 1) }
        if (self$o_loc %in% 22:24)         { o_move <- sample(c(0, -1,  1, -5), 1) }
        # middle
        if (self$o_loc %in% c(7:9, 12:14, 17:19)) {
          o_move <- sample(c(0, 1, -1, 5, -5), 1)
        }
        # increment
        self$o_loc <- self$o_loc + o_move

        # Create 1D grid
        if (self$moves == 29) {  # end of game

          self$overworld <- sample(c("\U1F384", "\U26C4", "\U1F328"), 25, TRUE)
          self$overworld[1] <- "\U1F43F️" # squirrel

          if (self$nuts < 8) {
            nut_diff <- 8 - self$nuts
            self$overworld[2] <- "\U1F480"  # skull
            self$overworld[3:(self$nuts + 2)] <- "\U1F330"  # nut
            self$overworld[(self$nuts + 3):((self$nuts + 2) + nut_diff)] <- "\U274C"  # cross
          } else {
            self$overworld[2] <- "\U1F4A4"  # zzz
            self$overworld[3:(self$nuts + 2)] <- "\U1F330"  # nut
          }

        } else {
          self$overworld <- rep("\U1F333", 25)
          self$overworld[self$n_loc] <- "\U1F330"
          self$overworld[self$s_loc] <- "\U1F43F️"
          self$overworld[self$o_loc] <- "\U1F989"
        }

        # Death emoji
        if (self$s_loc == self$o_loc) {
          self$overworld[self$o_loc] <- "\U1F480"  # skull
        }

        # Create 2D grid
        overworld_mat <- t(matrix(self$overworld, nrow = 5))

        # Clear console, print grid
        cat("\014")
        for (row in seq(nrow(overworld_mat))) {
          cat(overworld_mat[row, ], "\n")
        }

        # Increment move tally, print move and nut tallies
        self$moves <- self$moves + 1
        cat("Moves:", self$moves, "\nNuts:", self$nuts)

        # Owl attack routine
        if (self$s_loc == self$o_loc) {

          sonify::sonify(c(0, 2, 1), c(1, 1, 1), duration = 1)

          cat(
            "\nY O U   D I E D !",
            "\nThe owl ate you."
          )

          self$active <- FALSE  # change active state

        }

        # End game routine
        if (self$moves == 30) {

          if (self$nuts < 8) {  # lose

            sonify::sonify(c(0, 2, 1), c(1, 1, 1), duration = 1)

            cat(
              "\nY O U   D I E D !",
              "\nInsufficient winter nut cache!"
            )

          }

          if (self$nuts >= 8) {  # win

            sonify::sonify(c(0, 1), rep(1, 2), duration = 0.2)
            sonify::sonify(c(0, 1), rep(1, 2), duration = 0.2)
            sonify::sonify(c(0, 1), rep(1, 2), duration = 0.2)

            cat(
              "\nY O U   S U R V I V E D !",
              "\nSufficient winter nut cache!"
            )

          }

          self$active <- FALSE  # change active state

        }
      }

      # If game is over
      if (!self$active) {
        cat(
          "\nG A M E   O V E R",
          "\n* New game: x <- ActionSquirrel$new()",
          "\n* Source:   github.com/matt-dray/ActionSquirrel"
        )
      }

      invisible(self)

    }

  )
)
